File: others/WorksApplication/Worksap Package/workspace/China_Intern/China_Intern_Cpp/src/entities/abstract_dto_reader.cpp
#include "abstract_dto_reader.h"
#include "icsv_master_dto.h"
#include "../csv_parser/csv_reader.h"
#include "../constants/constants.h"
#include "../util/utilities.h"
#include <vector>
#include <string>
using namespace std;

vector<ICsvMasterDTO *> AbstractDTOReader::get_values() {
	return values;
}

void AbstractDTOReader::init() {
	ifstream file_reader(this->get_file_address().c_str());
	if (file_reader) {
		CSVReader csv_reader(file_reader, kDefaultCsvSeparator);
		vector<string> value;
		csv_reader.read_next(value); // skip header
		csv_reader.read_next(value);
		while (!value.empty()) {
			values.push_back(this->convert_array_to_dto(value));
			csv_reader.read_next(value);
		}
		Utilities::close_siently(csv_reader);
	}
	Utilities::close_siently(file_reader);
}

ICsvMasterDTO *AbstractDTOReader::convert_array_to_dto(vector<string> &value) {
	ICsvMasterDTO *dto = new ICsvMasterDTO();
	return dto;
}

string AbstractDTOReader::get_file_address() {
	string str = "csv address";
	return str;
}

AbstractDTOReader::~AbstractDTOReader() {
	for (vector<ICsvMasterDTO *>::iterator it = values.begin(); it != values.end(); it++) {
		delete *it;
	}
}

File: others/WorksApplication/Worksap Package/workspace/China_Intern/China_Intern_Java/src/jp/co/worksap/intern/entities/AbstractDTOReader.java
package jp.co.worksap.intern.entities;

import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import jp.co.worksap.intern.constants.Constants;
import jp.co.worksap.intern.constants.Messages;
import jp.co.worksap.intern.util.Utilities;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import au.com.bytecode.opencsv.CSVReader;

public abstract class AbstractDTOReader<E extends ICsvMasterDTO> {

	private List<E> values = null;
	protected Log log = null;

	protected AbstractDTOReader(String className) {
		log = LogFactory.getLog(className);
	}

	protected abstract E convertArrayToDTO(String[] value) throws IOException;

	/**
	 * file path of csv file
	 */
	protected abstract String getFileArress();

	public List<E> getValues() {
		return values;
	}

	protected void init() throws IOException {
		FileReader fileReader = null;
		try {
			fileReader = new FileReader(getFileArress());
			values = new ArrayList<E>();
			if (fileReader != null && fileReader.ready()) {
				CSVReader csvReader = new CSVReader(fileReader,
						Constants.DEFAULT_CSV_SEPARATOR);
				try {
					csvReader.readNext();
					// String[] titleLine = csvReader.readNext();
					String[] value = null;
					while ((value = csvReader.readNext()) != null) {
						if (value.length == 0){
							continue;
						}
						String[] newFields = new String[value.length];
						int i =0;
						for (String field : value){
							newFields[i++] = field.replace("\\n", "");
						}
						values.add(convertArrayToDTO(newFields));
					}
				} catch (IOException e) {
					log.error(Messages.ERROR_READ_CSV, e);
				} finally {
					Utilities.closeSilently(csvReader);
				}
			}
		} catch (FileNotFoundException e) {
			log.error(Messages.ERROR_CSV_NOT_FOUND, e);
		} finally {
			Utilities.closeSilently(fileReader);
		}
	}

}
